compiler/mir/mirgen.nim

```diff
@@ -263,163 +243,134 @@ proc typeOrVoid(c: TCtx, t: PType): PType {.inline.} =
-  ## Sets `info` as `n`'s origin information and appends the node to the
-  ## buffer. This operation is only allowed when all already buffered nodes
-  ## have had the origin information assigned too.
-  assert b.cursor == b.nodes.len
-  n.info = info
-  b.nodes.add n
-  inc b.cursor
-
-template add(f: var MirBuffer, n: MirNode) =
-  f.nodes.add n
-
-template subTree(f: var MirBuffer, n: MirNode, body: untyped) =
-  f.nodes.subTree n:
-    body
-
-func apply*(frag: var MirBuffer, id: SourceId) =
-  ## Associates all nodes added since the previous call to ``apply`` with the
-  ## origin information identified by `id`.
-  assert frag.nodes.len > frag.cursor
-  for i in frag.cursor..<frag.nodes.len:
-    frag.nodes[i].info = id
-
-  frag.cursor = frag.nodes.len
-
-func prepareForUse(sp: var SourceProvider): SourceId =
-  if sp.active.id.isNone:
-    sp.active.id = someOpt sp.map.add(sp.active.n)
+func isPure(tree: MirTree, n: NodePosition): bool =
+  ## Returns whether the expression at `n` is a pure expression.
+  case tree[n].kind
+  of mnkParam:
+    # sink parameters are mutable and thus not pure
+    tree[n].typ.kind != tySink
+  of mnkTemp, mnkConst, mnkLiteral, mnkProc, mnkType:
+    # note: during the translation phase, the name of a temporary is a pure
+    # expression
+    true
+  of mnkLocal:
+    # let bindings are pure, but only if they don't have a destructor (in
+    # which case they're movable)
+    tree[n].sym.kind in {skLet, skForVar} and not hasDestructor(tree[n].typ)
+  of mnkPathNamed, mnkPathPos:
+    isPure(tree, NodePosition tree.operand(n))
+  of mnkPathArray:
+    let arr = NodePosition tree.operand(n, 0)
+    case tree[arr].typ.skipTypes(abstractVar).kind
+    of tyArray:
+      # static array; pure when the array expression is pure and no index
+      # errors are possible
+      tree[n, 1].kind == mnkLiteral and isPure(tree, arr)
+    of tyUncheckedArray:
+      # no index errors are possible
+      isPure(tree, arr)
+    of tyString, tySequence, tyOpenArray, tyVarargs, tyCstring:
+      false # dynamically-sized arrays; index errors are possible
+    else:
+      unreachable(tree[arr].kind)
+  else:
+    false
 
-  result = sp.active.id[]
+func isStable(tree: MirTree, n: NodePosition): bool =
+  ## Returns whether the run-time address of the lvalue expression `n` is
+  ## always the same and whether the expression is side-effect free.
+  case tree[n].kind
+  of mnkConst, mnkGlobal, mnkParam, mnkLocal, mnkTemp, mnkAlias:
+    true
+  of mnkPathArray:
+    let arr = NodePosition tree.operand(n, 0)
+    case tree[arr].typ.skipTypes(abstractVar).kind
+    of tyArray:
+      # static arrays; stable when no index errors are possible
+      tree[n, 1].kind == mnkLiteral and isStable(tree, arr)
+    of tyUncheckedArray:
+      # cannot raise
+      isStable(tree, arr)
+    of tyString, tySequence, tyOpenArray, tyVarargs, tyCstring:
+      false # run-time arrays; could always raise
+    else:
+      unreachable()
```

Comment:
Isn't always `false` here a better choice? `unreachable` can help you discover paths to cover but since this is more or less an optimization(?) I think a "specialization" approach is better than trying to accurately pin-point all variants.

Reply:
In a sense it's an optimization, yeah. Nonetheless, from my experience, being exhaustive via `unreachable` forces one to think about all possible variations and also acts as another layer of defense against types/expressions reaching where they should not.

Thus, I don't think `false` is the better choice.